users/syncing.rst

```diff
@@ -79,6 +79,38 @@ synchronize the file. The block lists are compared to build a list of needed
 blocks, which are then requested from the network or copied locally, as
 described above.
 
+.. _case-sensitivity:
+
+Case Sensitivity in File Names
+------------------------------
+
+In principle Syncthing works with *case-sensitive* paths, meaning that
+`file.txt` and `FILE.txt` denote two independent things. It does not
+support synchronizing the contents of both as if they were somehow
+related.
+
+Some operating systems (e.g. Windows, Mac and Android) assume the
+opposite, treating them as the same file (or directory) and can never
+have both file names simultaneously. They will *preserve* the case
+from file creation, however.
+
+Thus, if two remote devices share these differing case variants with
+your local device, they cannot both be synchronized to such a system
+and will be reported as a case conflict by Syncthing. One of them will
+likely be pulled, and the other one marked with an appropriate error
+message. Which one "wins" is currently not predictable. In order to
+resolve such a case conflict situation, you need to decide on a
+consistent file name and **manually** enforce that across all
+involved, case-insensitive devices.
```

Comment:
I'd use a simpler scenario here, involving just two case-insensitive devices which happen to have a case-conflict. From support so far that seems the most likely scenario. With three the point about "winning" is unrelated to the problem at hand, and also not really a problem: In most scenarios you'd add one device later than the other, so the synced file will just be the one from the device added first.

And then there's also https://github.com/syncthing/syncthing/issues/6989, which means that for identical files no manual intervention will be necessary any more. That's probably the most common scenario with clusters of case-insensitive devices only. So then the relevant scenario would be case-sensitive with two files differing in case connected to case-insensitive.

I would also mention that the scenario will result in failed items and put the error message verbatim here, such that a search for that message may lead to this content.

Reply:
> I'd use a simpler scenario here, involving just two case-insensitive devices which happen to have a case-conflict. From support so far that seems the most likely scenario.

I wonder how people usually get into this situation. There must have been an unrelated synchronization / copying process before which screwed up file name case. Creating a file locally just before a remote device sends its own version, which it already had a name for before seeing our version? That's a typical conflict scenario anyway, because the contents would likely differ. Otherwise, what sequence of events would be needed to actually provoke a case conflict?

> With three the point about "winning" is unrelated to the problem at hand, and also not really a problem: In most scenarios you'd add one device later than the other, so the synced file will just be the one from the device added first.

They could both have been added long ago, but the local device was maybe just offline. When reconnecting to both conflicting (or not directly linked) remotes at the same time, the important point here is actually that the slightly quicker one will be reflected locally, while the error message from the other one might very well go unnoticed.

> And then there's also [syncthing/syncthing#6989](https://github.com/syncthing/syncthing/issues/6989), which means that for identical files no manual intervention will be necessary any more. That's probably the most common scenario with clusters of case-insensitive devices only. So then the relevant scenario would be case-sensitive with two files differing in case connected to case-insensitive.

Is any work being done on that? For me it's not clear from the issue what the automatic resolution would look like (which case variant will be chosen as "correct"). Basically what @calmh [said there](https://github.com/syncthing/syncthing/issues/6989#issuecomment-693992588), which also gives a rationale for my described three-device conflict above.

> I would also mention that the scenario will result in failed items and put the error message verbatim here, such that a search for that message may lead to this content.

I agree that the error message needs improvement, as it does not clearly point to the problem from a user's perspective. So let's follow up on this when a better message is implemented?

Reply:
> I wonder how people usually get into this situation. There must have been an unrelated synchronization / copying process before which screwed up file name case. Creating a file locally just before a remote device sends its own version, which it already had a name for before seeing our version? That's a typical conflict scenario anyway, because the contents would likely differ. Otherwise, what sequence of events would be needed to actually provoke a case conflict?

Someone mentioned using the same program on two devices, and that program apparently creates the same file sometimes in lower, sometimes upper case. In general I don't think much about "why" for case-insensitivity, that just creates headaches, I only think about how to prevent issues that exist (and plenty of weird stuff does exist).

> They could both have been added long ago, but the local device was maybe just offline. When reconnecting to both conflicting (or not directly linked) remotes at the same time, the important point here is actually that the slightly quicker one will be reflected locally, while the error message from the other one might very well go unnoticed.

There are scenarios where it's indeed situation where it's less clear which one wins, however my point is this has nothing to do with case conflicts. The same is true if those two devices modified the same file, producing a "real conflict". That's why I consider it unrelated information and thus I'd cut it.

> Is any work being done on that? For me it's not clear from the issue what the automatic resolution would look like (which case variant will be chosen as "correct"). Basically what @calmh [said there](https://github.com/syncthing/syncthing/issues/6989#issuecomment-693992588), which also gives a rationale for my described three-device conflict above.

I am planning to do it before the next release. I don't see how that comment is related to 3-way scenarios. As to how the automatic resolution would happen:

> We'll have to pick one casing essentially at random but the user apparently isn't picky anyhow or we wouldn't be here.

To make the choice deterministic (i.e. every device makes the same choice), I'd define that string `a` wins over `b`, if `a` > `b`.

Reply:
> Someone mentioned using the same program on two devices, and that program apparently creates the same file sometimes in lower, sometimes upper case. In general I don't think much about "why" for case-insensitivity, that just creates headaches, I only think about how to prevent issues that exist (and plenty of weird stuff does exist).

I'll try to master some of those headaches in search of a simpler example.

> There are scenarios where it's indeed situation where it's less clear which one wins, however my point is this has nothing to do with case conflicts. The same is true if those two devices modified the same file, producing a "real conflict". That's why I consider it unrelated information and thus I'd cut it.

Will case conflicts also result in the "losing" file being pulled with a `.case-conflict-...` suffix?  I assumed not, so the user will end up with only the winning version when they might have preferred the other.  That's the surprise I wanted to explain here.

> To make the choice deterministic (i.e. every device makes the same choice), I'd define that string `a` wins over `b`, if `a` > `b`.

This is definitely non-obvious, so I intend to incorporate it instead, when the implementation is ready. Can be a follow-up change though.

> > https://github.com/syncthing/syncthing/issues/6989#issuecomment-693992588 (emphasis added):
> > There should be some established / documented process for resolving a case conflict that doesn't fall under the above, i.e., when **we join two devices that have files with different contents and different name casing**. If one is older / less desired than the other, delete it. Otherwise rename one of them, etc.
> 
> I don't see how that comment is related to 3-way scenarios.

I thought he was talking about files with *same content*, and specifically about "joining" two *remote* devices.  Maybe just a misunderstanding.

Reply:
I am not interested in discussing these details any further, it requires too much time on to me almost completely irrelevant details - sorry for having engaged the discussion and now cutting it short on my side. I'll just answer the concrete question from the latest comment for completeness:

> Will case conflicts also result in the "losing" file being pulled with a `.case-conflict-...` suffix?

That's not happening now. It's on the table to maybe do it, but nothing is decided or planned on that.

For me to approve this PR, the "winning stuff" has to be removed. However that explicitly means that if anyone else is happy with leaving the "winning" stuff in there, I encourage that person to approve and merge.

Reply:
Sorry if I complicated this matter beyond what is necessary.  Details may seem irrelevant to you, with a deep background on how everything works.  Most users don't have that and will report bugs for "works as designed" behavior.

I'm okay with skipping further discussion.  I requested a review to make sure there are no factually **wrong** details in here, which I understood is the case.  Any correct documentation is better than none.

Reply:
> > Will case conflicts also result in the "losing" file being pulled with a `.case-conflict-...` suffix?
> 
> That's not happening now. It's on the table to maybe do it, but nothing is decided or planned on that.

Just a side note, but this would be really nice to have though, as in other conflicting cases Syncthing solves the conflicts automatically by renaming the files (or even overwrites them with `<maxConflicts>0</maxConflicts>`), but with a case conflict, the files get stuck waiting for a manual user intervention, which the user may not even be aware of, unless they open the GUI.

Having this would also make the "winning" problem not a big deal anymore, I guess.